samples/Twitter.gs

```diff
@@ -0,0 +1,113 @@
+    var response = UrlFetchApp.fetch(url, {
+      headers: {
+        Authorization: 'Bearer ' + service.getAccessToken()
+      },
+      muteHttpExceptions: true
+    });
+    var result = JSON.parse(response.getContentText());
+    Logger.log(JSON.stringify(result, null, 2));
+  } else {
+    var authorizationUrl = service.getAuthorizationUrl();
+    Logger.log('Open the following URL and re-run the script: %s',
+      authorizationUrl);
+  }
+}
+
+/**
+ * Reset the authorization state, so that it can be re-tested.
+ */
+function reset() {
+  getService().reset();
+  PropertiesService.getUserProperties().deleteProperty("code_challenge");
+  PropertiesService.getUserProperties().deleteProperty("code_verifier");
+}
+
+/**
+ * Configures the service.
+ */
+function getService() {
+  pkceChallengeVerifier();
+  var userProps = PropertiesService.getUserProperties();
+  return OAuth2.createService('Twitter')
+    // Set the endpoint URLs.
+    .setAuthorizationBaseUrl(
+      'https://twitter.com/i/oauth2/authorize?code_challenge_method=S256&code_challenge=' + userProps.getProperty("code_challenge"))
+    .setTokenUrl(
+      'https://api.twitter.com/2/oauth2/token?code_verifier=' + userProps.getProperty("code_verifier"))
+
+    // Set the client ID and secret.
+    .setClientId(CLIENT_ID)
+    .setClientSecret(CLIENT_SECRET)
+
+    // Set the name of the callback function that should be invoked to
+    // complete the OAuth flow.
+    .setCallbackFunction('authCallback')
+
+    // Set the property store where authorized tokens should be persisted.
+    .setPropertyStore(userProps)
+
+    // Set the scopes to request (space-separated for Twitter services).
+    .setScope('users.read tweet.read offline.access')
+
+    .setTokenHeaders({
+      'Authorization': 'Basic ' + Utilities.base64Encode(CLIENT_ID + ':' + CLIENT_SECRET),
+      'Content-Type': 'application/x-www-form-urlencoded'
+    })
+}
+
+/**
+ * Handles the OAuth callback.
+ */
+function authCallback(request) {
+  var service = getService();
+  var authorized = service.handleCallback(request);
+  if (authorized) {
+    return HtmlService.createHtmlOutput('Success!');
+  } else {
+    return HtmlService.createHtmlOutput('Denied.');
+  }
+}
+
+/**
+ * Logs the redict URI to register.
+ */
+function logRedirectUri() {
+  Logger.log(OAuth2.getRedirectUri());
+}
+
+/**
+ * Generates code_verifier & code_challenge for PKCE
+ */
+function pkceChallengeVerifier() {
+  var userProps = PropertiesService.getUserProperties();
+  if (!userProps.getProperty("code_verifier")) {
+    var verifier = "";
+    var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~";
+
+    for (var i = 0; i < 128; i++) {
+      verifier += possible.charAt(Math.floor(Math.random() * possible.length));
+    }
+
+    var sha256Hash = Utilities.computeDigest(Utilities.DigestAlgorithm.SHA_256, verifier)
+
+    var challenge = Utilities.base64Encode(sha256Hash)
```

Comment:
https://developers.google.com/apps-script/reference/utilities/utilities#base64EncodeWebSafe(Byte)

Reply:
@sqrrrl — tried `Utilities.base64EncodeWebSafe` but it still ended up creating an `=` at the end 😕

![Screenshot 2022-01-24 at 23 38 00](https://user-images.githubusercontent.com/37455462/150839639-b8d7ac24-60f4-4fdc-a423-3b53fa2deeb1.png)

Figured it might be best to adapt the snippet from [the official docs](https://www.oauth.com/oauth2-servers/pkce/authorization-request/) — 

![Screenshot 2022-01-24 at 23 40 02](https://user-images.githubusercontent.com/37455462/150839881-faa682af-f440-4291-a3f2-e8b5a82ebdd5.png)

You can try meddling with just the PKCE parts using [this script](https://script.google.com/d/1GSIco59T0ZPPRP3SNXQjZ0wYsoeWQiU8mpbrcusomndmWHSK3KyrVY78/edit?usp=sharing) — I wrote a [small post about](https://script.gs/generate-code-verifier-challenge-for-oauth2-with-pkce/) it too 😅

Reply:
Padding is allowed in websafe base64, but looks like twitter is particularly sensitive to it :(